common/http_test.cpp

```diff
@@ -0,0 +1,1217 @@
+	req2->SetHeader("X-WhichRequest", "2");
+	auto initiate_client2 = [&client1_response, &client2, &req2]() {
+		client2.AsyncCall(
+			req2,
+			[](http::ExpectedIncomingResponsePtr exp_resp) {
+				ASSERT_TRUE(exp_resp) << exp_resp.error().String();
+			},
+			[&client1_response](http::ExpectedIncomingResponsePtr exp_resp) {
+				ASSERT_TRUE(exp_resp) << exp_resp.error().String();
+				auto resp = exp_resp.value();
+
+				ASSERT_TRUE(resp->GetHeader("X-WhichResponse"));
+				EXPECT_EQ(resp->GetHeader("X-WhichResponse").value(), "2");
+
+				// Finish the first request.
+				ASSERT_TRUE(client1_response);
+				client1_response->AsyncReply([](error::Error err) { ASSERT_FALSE(err); });
+			});
+	};
+
+	http::ServerConfig server_config;
+	http::Server server(server_config, loop);
+	server.AsyncServeUrl(
+		"http://127.0.0.1:" TEST_PORT,
+		[](http::ExpectedIncomingRequestPtr exp_req) {
+			ASSERT_TRUE(exp_req) << exp_req.error().String();
+		},
+		[&client1_response, &initiate_client2](http::ExpectedIncomingRequestPtr exp_req) {
+			ASSERT_TRUE(exp_req) << exp_req.error().String();
+			auto req = exp_req.value();
+
+			ASSERT_TRUE(req->GetHeader("X-WhichRequest"));
+			if (req->GetHeader("X-WhichRequest").value() == "1") {
+				// Start the response, but don't complete it now.
+				auto exp_resp = req->MakeResponse();
+				ASSERT_TRUE(exp_resp) << exp_resp.error().String();
+				client1_response = exp_resp.value();
+				client1_response->SetStatusCodeAndMessage(200, "Success");
+				client1_response->SetHeader("X-WhichResponse", "1");
+
+				initiate_client2();
+			} else if (req->GetHeader("X-WhichRequest").value() == "2") {
+				// Complete this response.
+				auto exp_resp = req->MakeResponse();
+				ASSERT_TRUE(exp_resp) << exp_resp.error().String();
+				auto resp = exp_resp.value();
+
+				resp->SetStatusCodeAndMessage(200, "Success");
+				resp->SetHeader("X-WhichResponse", "2");
+				resp->AsyncReply([](error::Error err) { ASSERT_FALSE(err); });
+			} else {
+				FAIL() << "Unexpected X-WhichRequest header";
+			}
+		});
+
+	loop.Run();
+}
+
+class BodyOfXes : virtual public io::Reader {
+public:
+	expected::ExpectedSize Read(vector<uint8_t>::iterator start, vector<uint8_t>::iterator end) {
+		auto iter_count = end - start;
+		size_t read;
+		if (iter_count + count_ > TARGET_BODY_SIZE) {
+			read = TARGET_BODY_SIZE - count_;
+		} else {
+			read = iter_count;
+		}
+
+		for (size_t i = 0; i < read; i++) {
+			start[i] = TransferFunction(count_ + i);
+		}
+		count_ += read;
+		return read;
+	}
+
+	static uint8_t TransferFunction(size_t index) {
+		// Fill in a specific pattern to try to catch offset/block errors: Raise the input
+		// number to the power of 1.1 and round to the nearest integer. If it's odd, return
+		// 'X', if it's even, return 'x'. Due to the exponent, this pattern will change
+		// slightly throughout the sequence, giving us a chance to catch offset errors.
+		// (Note: Try printing it, the pattern is mesmerizing to watch!)
```

Comment:
Nice! Where does this come from?

Reply:
Nowhere, I invented it! But it is quite related to non-linear functions used in audio synthesizer/distortion effects (for guitars and such), which interest me. So I have a prior interest. 🙂

Reply:
Fun one!

Reply:
> TransferFunctions :)

Correct!